### ISSUE TITLE
[Radiance] Move away from codenames for individual modules?

### ISSUE THREAD

**Issue by kirill-grouchnikov** [kirill-grouchnikov/radiance#315]
If this ever happens, it's going to touch pretty much every single public API in Radiance.

Here's a wild idea. There's a lot of codenames for different modules under the Radiance umbrella. It started with Substance, then Flamingo that was a separate thing with additional Substance skinning, then Trident split off of Substance, then Neon was split to share common code, then Flamingo moved to depend on Substance. And along the way, there's been a bunch of tools that started with more "accessible" names, which then moved to something that I can't remember which one is which (Apollo? Zodiac? Electra? Hyperion)? And then there's Kotlin extension libraries with names that have nothing to do with the Radiance modules that they extend.

What if Trident was simply `animation`? Its Maven coordinates would be `org.pushing-pixels:radiance-animation`. Its APIs would be under `org.pushingpixels.radiance.animation` package. Substance would be `laf` with `org.pushing-pixels:radiance-laf` under the `org.pushingpixels.radiance.laf` package, with `SubstanceLookAndFeel` becoming `RadianceLookAndFeel`.

It'd certainly be a bit more cohesive to describe what Radiance is. At the expense of essentially breaking every single API - which would be 99% renaming of packages, classes and methods. But that's still probably well into thousands of such changes. I think it's worth it for the long run, and I don't hold much sentimental attachment to the existing codenames anymore.

**Comment by kirill-grouchnikov** [comment]
One possible way to make the migration just a tad easier would be to completely separate the "regular" Radiance releases that have new features and bug fixes from this big refactoring (that might, by the way, fold Neon back into Substance now that Flamingo requires Substance as a dependency). 

So, for example, release 4.0 (hypothetically) is a regular release at some point in H1.2020, and then straight away there's a 5.0 that only does the renaming without any features or bug fixes.

**Comment by enwired** [comment]
Doing the renaming as a separate release would certainly help the transition.  If you only change package names but keep class names the same, my IDE could probably easily fix the imports to the new packages.

**Comment by ragunathjawahar** [comment]
@enwired can you please elaborate on "my IDE could probably easily fix the imports"?

**Comment by kirill-grouchnikov** [comment]
In that still-hypothetical world, all APIs from `org.pushingpixels.trident.api` would move to be under `org.pushingpixels.radiance.animation` (including the relevant subpackage structure). Fixing the imports would be a search-and-replace within your IDE for the relevant top-level Radiance modules and their APIs.

**Comment by ragunathjawahar** [comment]
Okay, thank you for clarifying. In that situation, we could also use the [bash script](https://gist.github.com/dlew/5db1b780896bbc6f542e7c00a11db6a0) folks used to migrate from Android Support Libraries to AndroidX as long as a CSV mapping file for these classes are made available.

**Comment by kirill-grouchnikov** [comment]
Given a much smaller reach of this library, I do not intend to provide such a mapping file.

**Comment by ragunathjawahar** [comment]
I agree that mapping at a class level is not worth the effort for a smaller audience, but package mapping for all artifacts could help.

**Comment by kirill-grouchnikov** [comment]
Here's the plan. Release 4.5 has a smaller scope, with around a dozen bug fixes and API alignments. It will be released in a couple of weeks, around beginning of October.

Two weeks later, I'm planning to release version 5.0 which will be the big refactoring to rename all the modules / libraries / classes / APIs to follow the functional boundaries. More specifically:

- `org.pushingpixels.trident` -> `org.pushingpixels.radiance.animation`
- `org.pushingpixels.neon` -> `org.pushingpixels.radiance.common`
- `org.pushingpixels.substance` -> `org.pushingpixels.radiance.laf`
- `org.pushingpixels.flamingo` -> `org.pushingpixels.radiance.components`
- `org.pushingpixels.substance.extras` -> `org.pushingpixels.radiance.laf.extras`
- `org.pushingpixels.ember` -> `org.pushingpixels.radiance.laf.kxt`
- `org.pushingpixels.meteor` -> `org.pushingpixels.radiance.swing.kxt`
- `org.pushingpixels.plasma` -> `org.pushingpixels.radiance.flamingo.kxt`
- `org.pushingpixels.torch` -> `org.pushingpixels.radiance.animation.kxt`
- `org.pushingpixels.tools.apollo` -> `org.pushingpixels.radiance.tools.schemeeditor`
- `org.pushingpixels.tools.beacon` -> `org.pushingpixels.radiance.tools.lafdebugger`
- `org.pushingpixels.tools.electra` -> will be removed
- `org.pushingpixels.tools.hyperion` -> `org.pushingpixels.radiance.tools.shapereditor`
- `org.pushingpixels.tools.ignite` -> `org.pushingpixels.radiance.tools.gradleplugin.svgtranscoder`
- `org.pushingpixels.tools.lightbeam` -> `org.pushingpixels.radiance.tools.lafbenchmark`
- `org.pushingpixels.tools.photon` -> `org.pushingpixels.radiance.tools.svgtranscoder`
- `org.pushingpixels.tools.zodiac` -> `org.pushingpixels.radiance.tools.screenshot`

Classes that used codenames, such as `SubstanceLookAndFeel`, `TridentConfig` etc will be renamed to follow the functionality of the matching API sub-surface. For example:

- `SubstanceCortex` -> `RadianceLafCortex`
- `TridentCortex` -> `RadianceAnimationCortex`
- `SubstanceButtonUI` -> `RadianceButtonUI`

The names here are not final, and may be tweaked in the next few weeks for the final release 5.0

**Comment by kirill-grouchnikov** [comment]
5.0 release is scheduled for October 25th

**Comment by kirill-grouchnikov** [comment]
The new modules are live at https://repo1.maven.org/maven2/org/pushing-pixels/

It will take a couple more hours for them to be visible in search results on Maven Central
